doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,82 +1,181 @@
+        self.next = next
+    
+    def delete(self):
+        if self.prev:
+            self.next.prev = self.prev
+        if self.next:
+            self.prev.next = self.next
+            
+"""
+Our doubly-linked list class. It holds references to 
+the list's head and tail nodes.
+"""
+class DoublyLinkedList:
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+        self.length = 1 if node is not None else 0
+
+    def __len__(self):
+        return self.length
+    
+    """
+    Wraps the given value in a ListNode and inserts it 
+    as the new head of the list. Don't forget to handle 
+    the old head node's previous pointer accordingly.
+    """
+    def add_to_head(self, value):
+        #create new_node
+        new_node = ListNode(value)
+        if self.head is None: #adding to an empty list
+            self.head = new_node
+            self.tail = new_node
+        
+        else:#adding to non empty list
+             # head -> tail -> None    newNode 
+            new_node.next = self.head # newNode -> head
+            self.head.prev = new_node # newNode   <- head
+            self.head = new_node # head = newNode
+            # newHead -> oldHead -> tail -> None
+        self.length += 1
+        
+    """
+    Removes the List's current head node, making the
+    current head's next node the new head of the List.
+    Returns the value of the removed Node.
+    """
+    def remove_from_head(self):
+        old_head = self.head
+        if self.head == None:# remove head from empty list
+            return None   
+        if self.head is self.tail:
+            self.head = None
+            self.tail = None 
+        elif self.tail.prev is self.head:
+            old_head = self.tail.prev
+            self.tail.prev = None # head -> tail <-> None
+            self.head.next = None
+            self.head = self.tail
+        else:
+            self.head.next = None # head  tail <-> None
+        self.length -= 1
+        return old_head.value
+    """
+    Wraps the given value in a ListNode and inserts it 
+    as the new tail of the list. Don't forget to handle 
+    the old tail node's next pointer accordingly.
+    """
+    def add_to_tail(self, value):
+        new_node = ListNode(value)
+        if self.tail is None:# Add tail to empty list
+            self.head = new_node
+            self.tail = new_node
+        else:# Add tail to populated list 
+            # currHead <-> currTail <-> None    NewNode
+            # currTail -> newNode
+            new_node.prev = self.tail
+            self.tail.next = new_node  # 1 -> 30 # 1 <- 30
+            self.tail = new_node
+            # currTail <- newNode
+            # newNode = tail     
+        self.length += 1  
+    """
+    Removes the List's current tail node, making the 
+    current tail's previous node the new tail of the List.
+    Returns the value of the removed Node.
+    """
+    def remove_from_tail(self):

```

Comment:
A very similar logic will apply to `remove_from_tail` as it does to `remove_from_head`. Take a look at the comment above.